cep18/src/main.rs

```diff
@@ -0,0 +1,452 @@
+    write_total_supply_to(total_supply_uref, new_total_supply);
+    events::record_event_dictionary(Event::Burn(Burn { owner, amount }))
+}
+
+/// Initiates the contracts states. Only used by the installer call,
+/// later calls will cause it to revert.
+#[no_mangle]
+pub extern "C" fn init() {
+    if get_key(ALLOWANCES).is_some() {
+        revert(Cep18Error::AlreadyInitialized);
+    }
+    let package_hash = get_named_arg::<Key>(PACKAGE_HASH);
+    put_key(PACKAGE_HASH, package_hash);
+    storage::new_dictionary(ALLOWANCES).unwrap_or_revert();
+    let balances_uref = storage::new_dictionary(BALANCES).unwrap_or_revert();
+    let initial_supply = runtime::get_named_arg(TOTAL_SUPPLY);
+    let caller = get_caller();
+    write_balance_to(balances_uref, caller.into(), initial_supply);
+
+    let security_badges_dict = storage::new_dictionary(SECURITY_BADGES).unwrap_or_revert();
+    dictionary_put(
+        security_badges_dict,
+        &base64::encode(Key::from(get_caller()).to_bytes().unwrap_or_revert()),
+        SecurityBadge::Admin,
+    );
+
+    let admin_list: Option<Vec<Key>> =
+        utils::get_optional_named_arg_with_user_errors(ADMIN_LIST, Cep18Error::InvalidAdminList);
+    let minter_list: Option<Vec<Key>> =
+        utils::get_optional_named_arg_with_user_errors(MINTER_LIST, Cep18Error::InvalidMinterList);
+    let burner_list: Option<Vec<Key>> =
+        utils::get_optional_named_arg_with_user_errors(BURNER_LIST, Cep18Error::InvalidBurnerList);
+    let mint_and_burn_list: Option<Vec<Key>> = utils::get_optional_named_arg_with_user_errors(
+        MINT_AND_BURN_LIST,
+        Cep18Error::InvalidMintAndBurnList,
+    );
+
+    if let Some(minter_list) = minter_list {
+        for minter in minter_list {
+            dictionary_put(
+                security_badges_dict,
+                &base64::encode(minter.to_bytes().unwrap_or_revert()),
+                SecurityBadge::Minter,
+            );
+        }
+    }
+    if let Some(burner_list) = burner_list {
+        for burner in burner_list {
+            dictionary_put(
+                security_badges_dict,
+                &base64::encode(burner.to_bytes().unwrap_or_revert()),
+                SecurityBadge::Burner,
+            );
+        }
+    }
+    if let Some(mint_and_burn_list) = mint_and_burn_list {
+        for mint_and_burn in mint_and_burn_list {
+            dictionary_put(
+                security_badges_dict,
+                &base64::encode(mint_and_burn.to_bytes().unwrap_or_revert()),
+                SecurityBadge::MintAndBurn,
+            );
+        }
+    }
+    if let Some(admin_list) = admin_list {
+        for admin in admin_list {
+            dictionary_put(
+                security_badges_dict,
+                &base64::encode(admin.to_bytes().unwrap_or_revert()),
+                SecurityBadge::Admin,
+            );
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn change_security() {
```

Comment:
@gyroflaw @gRoussac an event on this method would be necessary too. WDYT?

Reply:
@davidatwhiletrue I agree, we should add it in to make it consistent across all other entrypoints